avr-hal-generic/src/adc.rs

```diff
@@ -0,0 +1,154 @@
+        pub use avr_hal::adc::*;
+
+        pub struct $Adc {
+            peripheral: $ADC,
+            reading_channel: Option<$ID>,
+        }
+
+        impl $Adc {
+            pub fn new(peripheral: $ADC, settings: AdcSettings) -> $Adc {
+                let s = Self { peripheral, reading_channel: None } ;
+                s.enable(settings);
+                s
+            }
+
+            fn enable(&self, settings: AdcSettings) {
+                self.peripheral.adcsra.write(|w| {
+                    w.aden().set_bit();
+                    match settings.clock_divider {
+                        ClockRateDivision::Factor2 => w.adps().prescaler_2(),
+                        ClockRateDivision::Factor4 => w.adps().prescaler_4(),
+                        ClockRateDivision::Factor8 => w.adps().prescaler_8(),
+                        ClockRateDivision::Factor16 => w.adps().prescaler_16(),
+                        ClockRateDivision::Factor32 => w.adps().prescaler_32(),
+                        ClockRateDivision::Factor64 => w.adps().prescaler_64(),
+                        ClockRateDivision::Factor128 => w.adps().prescaler_128(),
+                    }});
+                self.peripheral.admux.write(|w| match settings.ref_voltage {
+                    ReferenceVoltage::Aref => w.refs().aref(),
+                    ReferenceVoltage::AVcc => w.refs().avcc(),
+                    ReferenceVoltage::Internal => w.refs().internal(),
+                });
+                                                
+            }
+
+            fn disable(&self) {
+                self.peripheral.adcsra.reset();
+            }
+
+            pub fn release(self) -> $ADC {
+                self.disable();
+                self.peripheral
+            }
+        }
+
+        impl<WORD, PIN> OneShot<$Adc, WORD, PIN> for $Adc
+        where
+            WORD: From<u16>,
+            PIN: Channel<$Adc, ID=$ID>,
+        {
+            type Error = AdcError;
+
+            fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
+                match (self.reading_channel, self.peripheral.adcsra.read().adsc().bit_is_set()) {
+                    (Some(channel), true) if channel == PIN::channel() =>  Err(nb::Error::WouldBlock),
+                    (Some(channel), false) if channel == PIN::channel() => {
+                        self.reading_channel = None;
+                        Ok(self.peripheral.adc.read().bits().into())
+                    },
+                    (Some(_), _) => Err(nb::Error::Other(AdcError::AlreadyReaddingOtherChannel)),
+                    (None, _) => {
+                        self.reading_channel = Some(PIN::channel());
+                        self.peripheral.admux.modify(|_, w| w.mux().variant(PIN::channel()));
+                        self.peripheral.adcsra.modify(|_, w| w.adsc().set_bit());
+                        Err(nb::Error::WouldBlock)
+                    }
+                }
```

Comment:
I think the error when re-reading with a different pin is not the right choice.  If you don't want to support this case, I would `panic!()` as this would clearly be a programmer error and not an unforseeable hardware error.  But I would instead suggest to restart the measurement on the other pin when this happens.  Rough outline (untested of course):

```rust
match (self.reading_channel, self.peripheral.adcsra.read().adsc().bit_is_set()) {
    // Measurement on current pin is ongoing
    (Some(channel), true) if channel == PIN::channel() =>  Err(nb::Error::WouldBlock),
    // Measurement on current pin completed
    (Some(channel), false) if channel == PIN::channel() => {
        self.reading_channel = None;
        Ok(self.peripheral.adc.read().bits().into())
    },
    // Measurement on other pin is ongoing
    (_, true) => {
        // Abort the current measurement when requesting a different pin.
        self.reading_channel = None;
        Err(nb::Error::WouldBlock)
    },
    // Start measurement
    (None, false) => {
        self.reading_channel = Some(PIN::channel());
        self.peripheral.admux.modify(|_, w| w.mux().variant(PIN::channel()));
        self.peripheral.adcsra.modify(|_, w| w.adsc().set_bit());
        Err(nb::Error::WouldBlock)
    }
}
```

What do you think?

---

Looking at other HALs, I see that `read()` is implemented synchroneously in a lot of cases which means they don't have to deal with this at all.  An asynchroneous version is of course a lot nicer.

Reply:
Sound good to me